### PR TITLE
Fixing exception for Table formatting type mismatch

### LINKED ISSUE #2868: Incorrect exception in Table value formatting if type and format mismatch
In the following example:

```
In [4]: from astropy.table import Table

In [5]: t = Table()

In [6]: t['a'] = ['a','b','c']

In [7]: t['a'].format = "%9.5f"

In [8]: print(t)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-8-d6c4f5caf31e> in <module>()
----> 1 print(t)

/Volumes/Raptor/Library/Python/3.4/lib/python/site-packages/astropy-1.0.dev9579-py3.4-macosx-10.8-x86_64.egg/astropy/table/table.py in __unicode__(self)
    664 
    665     def __unicode__(self):
--> 666         lines, n_header = self.formatter._pformat_table(self)
    667         return '\n'.join(lines)
    668     if six.PY3:

/Volumes/Raptor/Library/Python/3.4/lib/python/site-packages/astropy-1.0.dev9579-py3.4-macosx-10.8-x86_64.egg/astropy/table/pprint.py in _pformat_table(self, table, max_lines, max_width, show_name, show_unit, html, tableid)
    296         for col in six.itervalues(table.columns):
    297             lines, n_header = self._pformat_col(col, max_lines, show_name,
--> 298                                                 show_unit)
    299             cols.append(lines)
    300 

/Volumes/Raptor/Library/Python/3.4/lib/python/site-packages/astropy-1.0.dev9579-py3.4-macosx-10.8-x86_64.egg/astropy/table/pprint.py in _pformat_col(self, col, max_lines, show_name, show_unit)
    148         """
    149         outs = {}  # Some values from _pformat_col_iter iterator that are needed here
--> 150         col_strs = list(self._pformat_col_iter(col, max_lines, show_name, show_unit, outs))
    151         col_width = max(len(x) for x in col_strs)
    152 

/Volumes/Raptor/Library/Python/3.4/lib/python/site-packages/astropy-1.0.dev9579-py3.4-macosx-10.8-x86_64.egg/astropy/table/pprint.py in _pformat_col_iter(self, col, max_lines, show_name, show_unit, outs)
    239                                   format_func(col.format, col[(i,) + multidim1]))
    240                 else:
--> 241                     col_str = format_func(col.format, col[i])
    242                 yield col_str
    243             elif i == i0:

/Volumes/Raptor/Library/Python/3.4/lib/python/site-packages/astropy-1.0.dev9579-py3.4-macosx-10.8-x86_64.egg/astropy/table/pprint.py in _auto_format_func(format_, val)
     62             except:
     63                 raise ValueError('Unable to parse format string {0}'
---> 64                                  .format(format_))
     65     _format_funcs[format_] = format_func
     66     return out

ValueError: Unable to parse format string %9.5f
```

The exception is misleading. The issue is that the value needs to be a float. The 'proper' exception for this in Python is:

```
In [9]: "%9.5f" % 'a'
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-9-55c5d51877aa> in <module>()
----> 1 "%9.5f" % 'a'

TypeError: a float is required
```

The current error makes it sound like the format is malformed.

### PR DESCRIPTION
**Pull request by bsipocz**
Rebased version of #3261 

Based on the comment https://github.com/astropy/astropy/pull/3261#issuecomment-132759746 I think that PR was good to go, just got merge conflicts.

fixes #2868

### CODE REVIEW

**Comment by bsipocz** [comment]
this is a weird one, apparently locally it's seemingly randomly fails with ``E       AttributeError: 'Column' object has no attribute '_name'`` or passes without issues.

**Comment by taldcroft** [comment]
@bsipocz - thanks for picking this up.  Please see comments:

https://github.com/astropy/astropy/pull/3261#issuecomment-132748187
https://github.com/astropy/astropy/pull/3261#r22369623
https://github.com/astropy/astropy/pull/3261#r22382401

File: astropy/table/column.py [hunk 1]
```diff
@@ -407,6 +410,42 @@ def name(self, val):
...
+        except AttributeError:  # nothing set at all yet
+            prev_format = None
+
+        self._format = format_string  # set new format string
+
+        try:
+            # test whether it formats without error exemplarily
+            self.pformat(max_lines=1)
+
+            if self.dtype.kind == 'O':
```
Comment (taldcroft) [inline astropy/table/column.py]: Even though this is formally correct to do the full pre-checking here, I think it is better to not have a very expensive operation.  One could end up formatting a billion lines just to set the`format` attribute, which is not a good thing.

**Comment by bsipocz** [comment]
@taldcroft - I've mostly addressed your comments after redid the PR rebase fixes. I think most of the original comments were already dealt with, but please give it another review when you have time. 

I'll add a changelog, too once the CI passed.

File: astropy/table/pprint.py [hunk 2]
```diff
@@ -133,8 +133,8 @@ def _auto_format_func(format_, val):
                     break
             else:
                 # None of the possible string functions passed muster.
-                raise ValueError('Unable to parse format string {0}'
-                                 .format(format_))
+                raise TypeError('unable to parse format string {0} for its '
```
Comment (taldcroft) [inline astropy/table/pprint.py]: Keep as `ValueError`.   This is consistent with Python behavior.  The problem is not the *type* of the column but the *value* of the format string.

File: astropy/table/pprint.py [hunk 3]
```diff
@@ -431,29 +433,38 @@ def _pformat_col_iter(self, col, max_lines, show_name, show_unit, outs,
...
-                        col_str = (format_func(col_format, col[(i,) + multidim0]) +
-                                  ' .. ' +
-                                  format_func(col_format, col[(i,) + multidim1]))
-                else:
-                    col_str = format_func(col_format, col[i])
-                yield col_str
+                try:
+                    yield format_col_str(idx)
+                except TypeError:
+                    raise TypeError(
```
Comment (taldcroft) [inline astropy/table/pprint.py]: Catch `ValueError` and raise `ValueError`.

**Comment by bsipocz** [comment]
@taldcroft - Done.

> Thanks for sticking with this!

No problem, this was a low hanging fruit for my occasional triaging of reducing the number of open issues.

**Comment by taldcroft** [comment]
@bsipocz - Sorry, I had not noticed before this was milestoned for 2.0.2, but I think it should be 3.0.  The changelog could be something like:
```
New feature:
Improved exception handling and error messages when column ``format`` 
attribute is incorrect for the column type.

API change: 
When setting the column ``format`` attribute the value is now immediately validated. 
Previously one could set to any value and it was only checked when actually formatting the
column.
```

**Comment by bsipocz** [comment]
Thanks @taldcroft! I meant to come back to this, but it slipped through...

**Comment by taldcroft** [comment]
> I meant to come back to this, but it slipped through...

I'm very familiar with this phenomenon.

**Pull request merged**